### PR TITLE
Add cache model serialization

### PR DESCRIPTION
**Pull request by yifan-flex**

### CODE REVIEW

File: flow360/component/simulation/operating_condition.py [hunk 1]
```diff
@@ -1,60 +1,76 @@
...
             material=Air(),
         )
+        state._cached = ThermalStateCache(altitude=altitude, temperature_offset=temperature_offset)
+
+        return state
 
     @property
-    def altitude(self) -> LengthType.Positive:
-        return self._altitude
+    def altitude(self) -> Optional[LengthType.Positive]:
```
Comment (benflexcompute) [inline flow360/component/simulation/operating_condition.py:51]: Is the optional required here?

File: flow360/component/simulation/operating_condition.py [hunk 2]
```diff
@@ -1,60 +1,76 @@
...
+        return self._cached.altitude
```
Comment (benflexcompute) [inline flow360/component/simulation/operating_condition.py:52]: Also would be nice if we tell the user that it was not given by user instead of return none?

File: flow360/component/simulation/operating_condition.py [hunk 3]
```diff
@@ -87,29 +103,41 @@ class AerospaceCondition(Flow360BaseModel):
...
-    @validate_arguments
     @classmethod
+    @pd.validate_call
     def from_mach(
         cls,
-        mach: PositiveFloat,
+        mach: pd.PositiveFloat,
         alpha: float = 0,
         beta: float = 0,
         atmosphere: ThermalState = ThermalState(),
```
Comment (benflexcompute) [inline flow360/component/simulation/operating_condition.py:113]: is atmosphere a proper aerospace term here? I am not familiar

File: flow360/component/simulation/operating_condition.py [hunk 4]
```diff
@@ -87,29 +103,41 @@ class AerospaceCondition(Flow360BaseModel):
...
     reference_velocity_magnitude: Optional[VelocityType.Positive] = None
 
-    @validate_arguments
     @classmethod
+    @pd.validate_call
     def from_mach(
         cls,
-        mach: PositiveFloat,
+        mach: pd.PositiveFloat,
         alpha: float = 0,
```
Comment (benflexcompute) [inline flow360/component/simulation/operating_condition.py:111]: lets use unit for angles. Could you add a TODO here?

File: flow360/component/simulation/unit_system.py [hunk 5]
```diff
@@ -574,22 +574,6 @@ class _TemperatureType(_DimensionedType):
     dim = u.dimensions.temperature
     dim_name = "temperature"
 
-    @classmethod
```
Comment (benflexcompute) [inline flow360/component/simulation/unit_system.py:577]: Why is this removed?

File: tests/simulation_framework/test_cache_model.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+
+class TempThermalStateCache(Flow360BaseModel):
+    altitude: Optional[LengthType.Positive] = None
+    temperature_offset: Optional[TemperatureType] = None
+
+
+class TempThermalState(CacheModelBase):
+    temperature: TemperatureType.Positive = pd.Field(288.15 * u.K, frozen=True)
```
Comment (benflexcompute) [inline tests/simulation_framework/test_cache_model.py:24]: Could you add a todo here? We can probably somehow give user warning that the give altitude (if any) will not be valid anymore if the altitude was used and user try to change these fields.

File: tests/simulation_framework/test_cache_model.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    )
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as temp_file:
+        temp_file_name = temp_file.name
+        operating_condition.to_file(temp_file_name)
+
+    try:
+        with open(temp_file_name) as fp:
+            model_dict = json.load(fp)
+            assert model_dict["some_value"] == 1230
+            assert model_dict["thermal_state"]["_cached"]
```
Comment (benflexcompute) [inline tests/simulation_framework/test_cache_model.py:67]: can you instead of assert here, check the dict is exactly the same as a handwritten one?

Comment (benflexcompute) [reply tests/simulation_framework/test_cache_model.py:67]: or you can compare the two objects. The original one and the one loaded from JSON. Similar to what we do with SimulationParam seralizer.

File: flow360/component/simulation/framework/cache_model_base.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import abc
+from typing import Any, Dict
+
+import pydantic as pd
+
+from flow360.component.simulation.framework.base_model import Flow360BaseModel
+
+
+class CacheModelBase(Flow360BaseModel, metaclass=abc.ABCMeta):
```
Comment (benflexcompute) [inline flow360/component/simulation/framework/cache_model_base.py:9]: CachedModelBase

File: flow360/component/simulation/framework/cache_model_base.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    def __init__(self, **data):
+        cached = data.pop("_cached", None)
+        super().__init__(**data)
+        if cached:
+            self._cached = self.__annotations__["_cached"].model_validate(cached)
```
Comment (benflexcompute) [inline flow360/component/simulation/framework/cache_model_base.py:14]: It is going to be interesting when the `model_validate` gives error. Maybe we should catch this internally?

File: flow360/component/simulation/operating_condition.py (hunk 1 above)
Comment (yifan-flex) [reply flow360/component/simulation/operating_condition.py:51]: optional is for returning none

File: flow360/component/simulation/operating_condition.py (hunk 2 above)
Comment (yifan-flex) [reply flow360/component/simulation/operating_condition.py:52]: I think it's ok. If they haven't specified they shouldn't expect a value

File: flow360/component/simulation/operating_condition.py (hunk 1 above)
Comment (benflexcompute) [reply flow360/component/simulation/operating_condition.py:51]: Right. If we want to return none.

File: flow360/component/simulation/operating_condition.py (hunk 3 above)
Comment (yifan-flex) [reply flow360/component/simulation/operating_condition.py:113]: This was Maciej's suggestion in one of the iterations. @maciej-flexcompute I think it makes sense but not 100% sure

File: flow360/component/simulation/unit_system.py (hunk 5 above)
Comment (yifan-flex) [reply flow360/component/simulation/unit_system.py:577]: We shouldn't have a positive constrains on unit types as we discussed. This can be used for temperature offset for example

File: flow360/component/simulation/operating_condition.py (hunk 2 above)
Comment (benflexcompute) [reply flow360/component/simulation/operating_condition.py:52]: Yes but it would be still better to let them know why it is none.

File: flow360/component/simulation/unit_system.py (hunk 5 above)
Comment (benflexcompute) [reply flow360/component/simulation/unit_system.py:577]: so did you add .Positive for existing TemperatureType?

File: flow360/component/simulation/operating_condition.py (hunk 3 above)
Comment (benflexcompute) [reply flow360/component/simulation/operating_condition.py:113]: Ok then let's keep this for now.

File: tests/simulation_framework/test_cache_model.py (hunk 7 above)
Comment (benflexcompute) [reply tests/simulation_framework/test_cache_model.py:67]: @pytest.mark.usefixtures("array_equality_override")

File: flow360/component/simulation/operating_condition.py (hunk 1 above)
Comment (yifan-flex) [reply flow360/component/simulation/operating_condition.py:51]: https://stackoverflow.com/questions/39429526/how-to-specify-nullable-return-type-with-type-hints
This post suggested using Optional

Comment (benflexcompute) [reply flow360/component/simulation/operating_condition.py:51]: Sure

**Review by benflexcompute — APPROVED** [review]
